### PR TITLE
[infoscanner][discs] Check .nomedia before stacking

### LINKED ISSUE #16877: .nomedia does not ignore Bluray (BDMV) movie folders
## Problem Description
Kodi Leia correctly ignores movie media folders with .nomedia in the movie folder - it correctly doesn't find these folders when scanning for new content. e.g. the movie Arbitrage is not found in Arbitrage/.nomedia

BUT, no matter where a .nomedia file is placed in a Bluray folder, scanning for new content finds the movie.  For example, scanning this folder will find the movie 8 Mile:

8 Mile/.nomedia
8 Mile/BDMV/.nomedia
8 Mile/BDMV/STREAM/.nomedia

## Expected Behavior
Placing a .nomedia file in ANY movie folder should result in Kodi ignoring media from that folder.
(Or if it is easier to recognise a .nomedia file anywhere else e.g. in the BDMV folder, that would be fine).

## Actual Behavior
No matter where the .nomedia file is placed in a folder with BDMV, media in the folder is found by Kodi when scanning.

### To Reproduce
Kodi Leia 18.1 and Windows 8.1
Create a Bluray folder with BDMV and STREAM subfolders
Move .nomedia into the main folder (or anywhere in the subfolders)
Scan the folder
The movie is (wrongly) found

### Debuglog

The debuglog can be found here:
[[kodi.log](https://github.com/xbmc/xbmc/files/3817012/kodi.log)](url)

I have 7,800 titles and use other Media players which are happy with TV Series and Movies in the same root folder - I'm trying to hide these from Kodi as Kodi uses copies of the TV BDMVs in a different TV Series root folder.

### Your Environment
Kodi 18.1 on Windows 8.1

### PR DESCRIPTION
**Pull request by enen92**
## Description
`items.Stack();` translates any directory that is a DVD or Bluray folder structure to the respective "manifest" path (e.g. `folder/` gets translated to `/folder/video_ts/video_ts.ifo`. This effectively bypasses any possible checks for `.nomedia` in the root folder. Kodi needs to check for nomedia before considering stacking directory entries (remove the inner folders than have a `.nomedia` file inside). I've done this in the VideoInfoScanner.cpp file to avoid spreading those checks elsewhere for components that shouldn't be responsible for this (e.g. could have been added in `GetOpticalMediaPath` instead)

## Motivation and context
Fix https://github.com/xbmc/xbmc/issues/16877

## How has this been tested?
Runtime tested on linux with dvds and blurays

## What is the effect on users?
Dvd folder structures or bluray folder structures are no longer scanned to the library if they contain a `.nomedia` file inside.

### CODE REVIEW

**Comment by howie-f** [comment]
it should work this way. you may want to take a look at `std::remove_if` though. see example here https://github.com/xbmc/xbmc/pull/19355 😀

**Comment by enen92** [comment]
Unfortunately that implies that the container should implement `erase` or remove items based on an iterator position which is not the case for our custom `CFileItemList` (it's not a vector implementation but rather some custom class)

**Review by howie-f — APPROVED** [review]
arg, i feel dumb. but just a tiny bit

File: xbmc/video/VideoInfoScanner.cpp [hunk 1]
```diff
@@ -298,6 +298,11 @@ namespace VIDEO
       { // need to fetch the folder
         CDirectory::GetDirectory(strDirectory, items, CServiceBroker::GetFileExtensionProvider().GetVideoExtensions(),
                                  DIR_FLAG_DEFAULTS);
+        // do not consider inner folders with .nomedia
+        items.erase(std::remove_if(items.begin(), items.end(),
+                                   [&](const CFileItemPtr& item)
```
Comment (howie-f) [inline xbmc/video/VideoInfoScanner.cpp:303]: enough to capture `[this]` for the HasNoMedia() call.

**Review by howie-f — COMMENTED** [review]
one last nitpick, after that it's fine

File: xbmc/video/VideoInfoScanner.cpp (hunk 1 above)
Comment (enen92) [reply xbmc/video/VideoInfoScanner.cpp:303]: I assume that passes this by reference right?

Comment (howie-f) [reply xbmc/video/VideoInfoScanner.cpp:303]: no by val, but it‘s only a pointer and you can‘t change the value of `this` anyway

**Review by garbear — APPROVED** [review]
Fix looks correct

**Comment by enen92** [comment]
jenkins build and merge

**Comment by enen92** [comment]
jenkins build and merge

**Pull request merged**